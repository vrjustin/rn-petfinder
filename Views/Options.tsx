import React from 'react';
import {useSelector, useDispatch} from 'react-redux';
import {
  Text,
  View,
  TextInput,
  StyleSheet,
  TouchableOpacity,
} from 'react-native';
import {
  selectSearchParameters,
  setSearchParameters,
} from '../reducers/searchParamsReducer';

const Options: React.FC = () => {
  const dispatch = useDispatch();
  const searchParameters = useSelector(selectSearchParameters);
  const {distance, location, tagsPreferred} = searchParameters;

  const handleZipCodeChange = (newZip: string) => {
    dispatch(
      setSearchParameters({...searchParameters, location: {zipCode: newZip}}),
    );
  };

  const handleDistanceChange = (newDistance: string) => {
    let distanceValue =
      newDistance.trim() !== '' ? parseInt(newDistance, 10) : 1;
    distanceValue = Math.max(1, Math.min(500, distanceValue));
    dispatch(
      setSearchParameters({
        ...searchParameters,
        distance: distanceValue,
      }),
    );
  };

  const handleTagPress = (tag: string) => {
    if (!tagsPreferred.includes(tag)) {
      dispatch(
        setSearchParameters({
          ...searchParameters,
          tagsPreferred: [...tagsPreferred, tag],
        }),
      );
    } else {
      dispatch(
        setSearchParameters({
          ...searchParameters,
          tagsPreferred: tagsPreferred.filter(t => t !== tag),
        }),
      );
    }
  };

  return (
    <View style={styles.container}>
      <Text style={styles.label}>Enter ZIP Code:</Text>
      <TextInput
        style={styles.input}
        value={location.zipCode}
        onChangeText={handleZipCodeChange}
        keyboardType="numeric"
      />
      <Text style={styles.label}>Enter Search Distance:</Text>
      <TextInput
        style={styles.input}
        value={distance.toString()}
        onChangeText={handleDistanceChange}
        keyboardType="numeric"
      />
      {tagsPreferred.length > 0 && (
        <>
          <Text style={styles.label}>Preferred Tags:</Text>
          <View style={{flexDirection: 'row', flexWrap: 'wrap', padding: 8}}>
            {tagsPreferred.map((tag, index) => (
              <TouchableOpacity key={index} onPress={() => handleTagPress(tag)}>
                <View
                  key={index}
                  style={
                    !tagsPreferred.includes(tag) ? styles.tag : styles.activeTag
                  }>
                  <Text style={{color: 'white'}}>{tag}</Text>
                </View>
              </TouchableOpacity>
            ))}
          </View>
        </>
      )}
    </View>
  );
};

const styles = StyleSheet.create({
  container: {
    backgroundColor: '#f9f9f9',
    padding: 16,
    borderRadius: 8,
    marginVertical: 8,
  },
  label: {
    fontSize: 18,
    marginBottom: 8,
    color: '#333',
  },
  input: {
    borderWidth: 1,
    borderColor: '#ccc',
    borderRadius: 5,
    padding: 8,
    marginBottom: 16,
    fontSize: 16,
    backgroundColor: '#fff',
  },
  tag: {
    backgroundColor: 'gray',
    paddingVertical: 4,
    paddingHorizontal: 8,
    borderRadius: 8,
    marginRight: 8,
    marginBottom: 8,
  },
  activeTag: {
    backgroundColor: 'red',
    paddingVertical: 4,
    paddingHorizontal: 8,
    borderRadius: 8,
    marginRight: 8,
    marginBottom: 8,
  },
});

export default Options;
