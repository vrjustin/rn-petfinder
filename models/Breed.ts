interface Breed {
  name: string;
  _links: {
    type: {
      href: string;
    };
  };
}

export default Breed;
