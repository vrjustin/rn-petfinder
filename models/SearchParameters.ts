interface SearchParameters {
  location: {zipCode: string};
  distance: number;
}

export default SearchParameters;
