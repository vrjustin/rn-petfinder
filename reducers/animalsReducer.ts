import {createSlice, PayloadAction} from '@reduxjs/toolkit';
import {RootState} from '../stores/store';
import Animal from '../models/Animal';

interface AnimalsState {
  animals: Animal[];
  favorites: Animal[];
}

const initialState: AnimalsState = {
  animals: [],
  favorites: [],
};

const animalsSlice = createSlice({
  name: 'animals',
  initialState,
  reducers: {
    setAnimals(state, action: PayloadAction<Animal[]>) {
      state.animals = action.payload;
    },
    toggleFavorite(state, action: PayloadAction<Animal>) {
      const isFavorite = state.favorites.some(
        fav => fav.id === action.payload.id,
      );
      if (isFavorite) {
        state.favorites = state.favorites.filter(
          fav => fav.id !== action.payload.id,
        );
      } else {
        state.favorites.push(action.payload);
      }
    },
  },
});

export const selectAnimals = (state: RootState) => state.animals.animals;
export const selectFavorites = (state: RootState) => state.animals.favorites;

export const {setAnimals, toggleFavorite} = animalsSlice.actions;
export default animalsSlice.reducer;
