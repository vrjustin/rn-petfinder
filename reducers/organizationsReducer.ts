import {createSlice, PayloadAction} from '@reduxjs/toolkit';
import {RootState} from '../stores/store';
import Organization from '../models/Organization';

interface OrganizationsState {
  organizations: Organization[];
}

const initialState: OrganizationsState = {
  organizations: [],
};

const organizationsSlice = createSlice({
  name: 'organizations',
  initialState,
  reducers: {
    setOrganizations(state, action: PayloadAction<Organization[]>) {
      state.organizations = action.payload;
    },
  },
});

export const selectOrganizations = (state: RootState) =>
  state.organizations.organizations;

export const {setOrganizations} = organizationsSlice.actions;
export default organizationsSlice.reducer;
