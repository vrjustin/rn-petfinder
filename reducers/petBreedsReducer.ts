import {createSlice, PayloadAction} from '@reduxjs/toolkit';
import Breed from '../models/Breed';
import {RootState} from '../stores/store';

interface BreedsState {
  breeds: Breed[];
}

const initialState: BreedsState = {
  breeds: [],
};

const breedsSlice = createSlice({
  name: 'breeds',
  initialState,
  reducers: {
    setBreeds(state, action: PayloadAction<Breed[]>) {
      state.breeds = action.payload;
    },
  },
});

export const selectPetBreeds = (state: RootState) => state.petBreeds.breeds;

export const {setBreeds} = breedsSlice.actions;
export default breedsSlice.reducer;
