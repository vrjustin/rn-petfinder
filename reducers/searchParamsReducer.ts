import {createSlice, PayloadAction} from '@reduxjs/toolkit';
import {RootState} from '../stores/store';
import SearchParameters from '../models/SearchParameters';

interface SearchParametersState {
  searchParameters: SearchParameters;
}

const initialState: SearchParametersState = {
  searchParameters: {
    location: {
      zipCode: '90210',
    },
    distance: 500, //in miles
    tagsPreferred: [],
  },
};

const searchParametersSlice = createSlice({
  name: 'searchParameters',
  initialState,
  reducers: {
    setSearchParameters(state, action: PayloadAction<SearchParameters>) {
      state.searchParameters = action.payload;
    },
  },
});

export const selectSearchParameters = (state: RootState) =>
  state.searchParameters.searchParameters;

export const {setSearchParameters} = searchParametersSlice.actions;
export default searchParametersSlice.reducer;
