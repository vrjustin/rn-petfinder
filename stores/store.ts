import {combineReducers, configureStore} from '@reduxjs/toolkit';
import petTypesReducer from '../reducers/petTypesReducer';
import petBreedsReducer from '../reducers/petBreedsReducer';
import animalsReducer from '../reducers/animalsReducer';
import searchParamsReducer from '../reducers/searchParamsReducer';

const createDebugger = require('redux-flipper').default;

const rootReducer = combineReducers({
  petTypes: petTypesReducer,
  petBreeds: petBreedsReducer,
  animals: animalsReducer,
  searchParameters: searchParamsReducer,
});

export type RootState = ReturnType<typeof rootReducer>;

const store = configureStore({
  reducer: rootReducer,
  middleware: getDefaultMiddleware =>
    __DEV__
      ? getDefaultMiddleware({serializableCheck: false}).concat(
          createDebugger(),
        )
      : getDefaultMiddleware({serializableCheck: false}),
});

export default store;
