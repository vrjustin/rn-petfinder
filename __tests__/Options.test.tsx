import React from 'react';
import renderer from 'react-test-renderer';
import {Provider} from 'react-redux';
import * as reactRedux from 'react-redux';
import store from '../stores/store';
import SearchParameters from '../models/SearchParameters';
import * as searchParamsReducer from '../reducers/searchParamsReducer';
import Breed from '../models/Breed';
import Options from '../Views/Options';

const mockBreed: Breed = {
  name: 'MockBreed',
  _links: {
    type: {
      href: 'href',
    },
  },
};
const mockSearchParameters: SearchParameters = {
  location: {
    zipCode: '90210',
  },
  distance: 5,
  tagsPreferred: ['Tag1', 'Tag2', 'Tag3'],
  breedsPreferred: [mockBreed],
};

let renderedOptionsTree: any;

describe('Options', () => {
  afterEach(() => {
    jest.clearAllMocks();
  });

  it('renders correctly', () => {
    renderedOptionsTree = renderer.create(
      <Provider store={store}>
        <Options route={{params: {from: 'hello'}}} />
      </Provider>,
    );
    expect(renderedOptionsTree.toJSON()).toMatchSnapshot();
  });

  it('renders correctly when preferredTags and breeds are present', () => {
    jest
      .spyOn(searchParamsReducer, 'selectSearchParameters')
      .mockReturnValue(mockSearchParameters);
    renderedOptionsTree = renderer.create(
      <Provider store={store}>
        <Options route={{params: {from: 'hello'}}} />
      </Provider>,
    );
    expect(renderedOptionsTree.toJSON()).toMatchSnapshot();
  });

  it('updates displayZip state correctly on zip code change', () => {
    renderedOptionsTree = renderer.create(
      <Provider store={store}>
        <Options route={{params: {from: 'hello'}}} />
      </Provider>,
    );
    const textInput = renderedOptionsTree.root.findByProps({
      testID: 'zipCodeInput',
    });
    textInput.props.onChangeText('12345');

    const postChangeText = textInput.props.value;
    expect(postChangeText).toEqual('12345');
  });

  it('updates location zipCode state correctly on zip code blur', () => {
    const mockDispatch = jest.fn();
    jest.spyOn(reactRedux, 'useDispatch').mockReturnValue(mockDispatch);

    renderedOptionsTree = renderer.create(
      <Provider store={store}>
        <Options route={{params: {from: 'hello'}}} />
      </Provider>,
    );
    const textInput = renderedOptionsTree.root.findByProps({
      testID: 'zipCodeInput',
    });
    textInput.props.onChangeText('54321');
    textInput.props.onBlur();

    const dispatchedAction = mockDispatch.mock.calls[0][0];

    expect(dispatchedAction.type).toBe('searchParameters/setSearchParameters');
    expect(dispatchedAction.payload.location.zipCode).toEqual('54321');
  });

  it('updates distance state correctly on distance blur', () => {
    const mockDispatch = jest.fn();
    jest.spyOn(reactRedux, 'useDispatch').mockReturnValue(mockDispatch);

    renderedOptionsTree = renderer.create(
      <Provider store={store}>
        <Options route={{params: {from: 'hello'}}} />
      </Provider>,
    );
    const textInput = renderedOptionsTree.root.findByProps({
      testID: 'distanceInput',
    });
    textInput.props.onChangeText('55');
    textInput.props.onBlur();

    const dispatchedAction = mockDispatch.mock.calls[0][0];

    expect(dispatchedAction.type).toBe('searchParameters/setSearchParameters');
    expect(dispatchedAction.payload.distance).toEqual(55);
  });

  it('updates displayDistance state correctly on zip code change', () => {
    renderedOptionsTree = renderer.create(
      <Provider store={store}>
        <Options route={{params: {from: 'hello'}}} />
      </Provider>,
    );
    const textInput = renderedOptionsTree.root.findByProps({
      testID: 'distanceInput',
    });
    textInput.props.onChangeText('2');

    const postChangeText = textInput.props.value;
    expect(postChangeText).toEqual('2');
  });

  it('handles the preferred breeds tag press properly and removes that tag', () => {
    jest
      .spyOn(searchParamsReducer, 'selectSearchParameters')
      .mockReturnValue(mockSearchParameters);

    const mockDispatch = jest.fn();
    jest.spyOn(reactRedux, 'useDispatch').mockReturnValue(mockDispatch);

    renderedOptionsTree = renderer.create(
      <Provider store={store}>
        <Options route={{params: {from: 'petTypes'}}} />
      </Provider>,
    );

    const preferredBreedButton = renderedOptionsTree.root.findByProps({
      testID: 'breedsTagButton',
    });
    preferredBreedButton.props.onPress();

    // Get the latest dispatched action
    const dispatchedAction = mockDispatch.mock.calls[0][0];

    expect(dispatchedAction.type).toBe('searchParameters/setSearchParameters');
    expect(dispatchedAction.payload.breedsPreferred).toEqual([]);
  });

  it('handles the tags tag press properly and removes that tag', () => {
    jest
      .spyOn(searchParamsReducer, 'selectSearchParameters')
      .mockReturnValue(mockSearchParameters);

    const mockDispatch = jest.fn();
    jest.spyOn(reactRedux, 'useDispatch').mockReturnValue(mockDispatch);

    renderedOptionsTree = renderer.create(
      <Provider store={store}>
        <Options route={{params: {from: 'petTypes'}}} />
      </Provider>,
    );

    const tagButtons = renderedOptionsTree.root.findAllByProps({
      testID: 'tagsTagButton',
    });
    const firstTag = tagButtons[0];
    firstTag.props.onPress();

    // Get the latest dispatched action
    const dispatchedAction = mockDispatch.mock.calls[0][0];

    expect(dispatchedAction.type).toBe('searchParameters/setSearchParameters');
    expect(dispatchedAction.payload.tagsPreferred).toEqual(['Tag2', 'Tag3']);
  });
});
