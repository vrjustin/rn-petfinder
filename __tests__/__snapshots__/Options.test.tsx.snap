// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Options renders correctly 1`] = `
<View
  style={
    {
      "backgroundColor": "#f9f9f9",
      "borderRadius": 8,
      "marginVertical": 8,
      "padding": 16,
    }
  }
>
  <Text
    style={
      {
        "color": "#333",
        "fontSize": 18,
        "marginBottom": 8,
      }
    }
  >
    Enter ZIP Code:
  </Text>
  <TextInput
    keyboardType="numeric"
    onChangeText={[Function]}
    style={
      {
        "backgroundColor": "#fff",
        "borderColor": "#ccc",
        "borderRadius": 5,
        "borderWidth": 1,
        "fontSize": 16,
        "marginBottom": 16,
        "padding": 8,
      }
    }
    value="90210"
  />
  <Text
    style={
      {
        "color": "#333",
        "fontSize": 18,
        "marginBottom": 8,
      }
    }
  >
    Enter Search Distance:
  </Text>
  <TextInput
    keyboardType="numeric"
    onChangeText={[Function]}
    style={
      {
        "backgroundColor": "#fff",
        "borderColor": "#ccc",
        "borderRadius": 5,
        "borderWidth": 1,
        "fontSize": 16,
        "marginBottom": 16,
        "padding": 8,
      }
    }
    value="500"
  />
</View>
`;
